src/proto.js

```diff
@@ -242,6 +243,24 @@ exports.request = function (method, path) {
     return fn(err, res)
   }
 
+  // Captures the current async context by creating an async resource for the
+  // superagent request.
+  //
+  // This is necessary because async_hooks does not follow "Thenable" types,
+  // depsite promises being compatible with them. The "then" method is invoked
+  // asynchronously on a "nextTick", where the async context that originally
+  // created the "Thenable" object has been lost.
+  //
+  // By creating the async resource here and overriding the "then" method, we
+  // bind the request object to the async execution context, and can restore
+  // it via a call to runInAsyncScope in the overriden "then" function.
+  const asyncResource = new AsyncResource('com.segment.integration.Request')
+  const thenFunction = Reflect.get(req, 'then')
+  const thenAsyncHook = function then (resolve, reject) {
```

Comment:
Interesting. Is the current implementation different than:
```suggestion
  function thenAsyncHook (resolve, reject) {
```

Reply:
I think in the current code the function will be named `then`, so it leaks less details about the hack we're doing.

```js
const thenAsyncHook = function then () {}
console.log(thenAsyncHook.name)
console.log((function thenAsyncHook () {}).name)
```

```
$ node test.js
then
thenAsyncHook
```

Reply:
Interesting thanks!